### PR TITLE
api: railjson V3

### LINKED ISSUE #1909: Find a new design to split functional and cosmetic fields
## Description

Design a new railjson format that have cosmetic and functional fields split.

## Acceptance criteria

- [x] Cosmetic fields shouldn't depends on functionality (avoid sync)
- [x] Cosmetic fields should be optional
- [x] The front should be able to show something as well
- [x] This will result in a new version of railjson
- [x] Infra generation should be adapted to generate using new format
- [x] Editoast should work properly without cosmetic fields
- [x] Chartos should work properly without cosmetic fields
- [x] Core should work properly without cosmetic fields

## Definition of ready

- [ ] The enabler story is well defined
- [ ] Acceptance criterias are defined
- [ ] The individual responsible for validating the enbaler story is assigned
- [ ] Does not depend on an missing feature
- [ ] Negociable with the implementation team
- [ ] Valuable to the end users
- [ ] Estimable by the implementation team
- [ ] As small as it can be
- [ ] Testable by the implementation team

## Definition of done

- [ ] Acceptance criteria are met
- [ ] Automated tests are written against acceptance criteria
- [ ] Story ok-ed by Product Owner
- [ ] Documentation updated

### PR DESCRIPTION
**Pull request by Tguisnet**
close #1909

# Changelog in railjson v3

In a nutshell :

Appart from the errors, `ObjectRef` dissappeared from all objects to become a `String`.
(or a `Waypoint` for route entry/exit point)
Cosmetic fields are put in a struct for their scope (`SncfExtension`, or `IdentifierExtension` for example).
Those structs are put in an other struct containing all different scope of extension (`TrackSectionExtension`, or `OperationalPointExtension` for example) and they are optional.

### Tracksection :

- `navigability` : suppressed
- `extensions` : new field containing `TrackSectionExtensions` struct
- `TrackSectionExtensions` : new struct containing optional `TrackSectionSncfExtension` struct
- fields `line_code`, `line_name`, `track_number`, `track_name` are moved in `TrackSectionSncfExtension`

### Detector :

- `track` : type changed from `ObjectRef` to `String` (of the track's id)

### BufferStop :

- `track` : type changed from `ObjectRef` to `String` (of the track's id)

### Waypoint (new) :

- this is a polymorphic reference on `BufferStop` or `Detector`, This is used for the entry and exit point of the routes.

### ApplicableDirectionsTrackRange :

- `track` : type changed from `ObjectRef` to `String` (of the track's id)

### DirectionalTrackRange :

- `track` : type changed from `ObjectRef` to `String` (of the track's id)

### TrackEndpoint :

- `track` : type changed from `ObjectRef` to `String` (of the track's id)

### OperationalPoint :

- `extensions` : new field containing `OperationalPointExtensions` structure
-
- `OperationalPointExtensions` : new struct containing `OperationalPointSncfExtension` and `OperationalPointIdentifierExtension` structures both optional
- fields `name` and `uic` are moved in `OperationalPointIdentifierExtension`
- fields `ch_short_label`, `ch_long_label`, `ci`, `ch`, `trigram` are moved in `OperationalPointSncfExtension`
- `OperationalPointPart` : `track` type changed from `ObjectRef` to `String` (of the track's id)

### Route :

`entry_point` : type changed from `ObjectRef` to `Waypoint`
`exit_point` : type changed from `ObjectRef` to `Waypoint`
`release_detectors` : type change from `List<ObjectRef>` to `List<String>`

### Signal :

- `track` : type changed from `ObjectRef` to `String` (of the track's id)
- `linked_detector` : type changed from `Optional<ObjectRef>` an optional `Optional<String>`
- `extensions` : new field containing `SignalExtensions` struct
- `SignalExtensions` : new struct containing optional `SignalSncfExtension`
- fields `default_aspect`, `installation_type`, `is_in_service`, `is_lightable`, `is_operational`, `label`, `physical_organization_group`, `responsible_group`, `side`, `support_type`, `type_code`, `value` are moved in `SignalSncfExtension`

### Switch :

- `switch_type` : type changed from `ObjectRef` to `String`

### SwitchType :

- `bidirectional` : suppressed

### TrackSectionLink :

- `navigabilty` : suppressed

### CODE REVIEW

**Comment by multun** [comment]
is it what you had in mind?
```json
{
    "id": "track.42",
    "extensions": {
        "sncf": {"foo": "bar"}
    }
}
```

If so, metaprogramming feels overkill

**Comment by flomonster** [comment]
It's exactly what we want to achieve here. Metaprog kind of simplified a lot the code.

File: api/osrd_infra/schemas/infra.py [hunk 1]
```diff
@@ -188,12 +190,6 @@ class OperationalPoint(BaseObjectTrait):
 
     parts: List[OperationalPointPart]
     name: constr(max_length=255) = Field(description="Name of the operational point")
```
Comment (flomonster) [inline api/osrd_infra/schemas/infra.py]: Add it to an extension

File: api/osrd_infra/schemas/infra.py [hunk 2]
```diff
@@ -439,5 +412,90 @@ class RailJsonInfra(BaseModel):
...
+class TrackSectionSncfExtension(BaseModel):
+    line_code: int = Field(description="Code of the line used by the corresponding track section")
+    line_name: constr(max_length=255) = Field(description="Name of the line used by the corresponding track section")
+    track_number: int = Field(description="Number corresponding to the track used", ge=0)
+    track_name: constr(max_length=255) = Field(description="Name corresponding to the track used")
+
+
+@register_extension(object=OperationalPoint, name="sncf")
+class OperationalPointSncfExtension(BaseModel):
+    uic: int
```
Comment (flomonster) [inline api/osrd_infra/schemas/infra.py]: Move it to another extension

**Review by flomonster — CHANGES_REQUESTED** [review]
- [x] Remove `applicable_train_type` (not used in core)
- [ ] API could adapt its endpoint to download railjson without extensions
- [x] Rework refs. Could be less flexible.
- [ ] Rework signals cosmetic fields
- [x] Remove `bidirectional` from the SwitchPortConnection
  - The navigability is manged only with routes (needed to be documented)

**Comment by flomonster** [comment]
New route schema suggestion:

```python
class Route(BaseObjectTrait):
    """
    This class is used to describe routes on the infrastructure.
    """

    entry_point: ObjectReference = Field(description="Identifier and type used to define an entry point for the route")
    entry_point_direction: ObjectReference = Field(description="Identifier and type used to define an entry point for the route")
    exit_point: ObjectReference = Field(description="Identifier and type used to define an exit point for the route")
    release_detectors: List[ObjectReference] = Field(
        description="Detector allowing the release of resources reserved from the beginning of the route until this one"
    )
    switches_directions: List[Tuple[ObjectReference, str]] = ...
```

File: api/osrd_infra/schemas/infra.py [hunk 3]
```diff
@@ -406,6 +374,9 @@ class BufferStop(BaseObjectTrait, TrackLocationTrait):
 
     applicable_directions: ApplicableDirections = Field(description="Direction of the application of the buffer stop")
 
+    def ref(self):
+        return DetectorReference(type="Detector", id=self.id)
```
Comment (Castavo) [inline api/osrd_infra/schemas/infra.py]: There must be something I have misunderstood, why is it not :
```suggestion
        return BufferStopReference(type="BufferStop", id=self.id)
```

File: api/osrd_infra/schemas/infra.py [hunk 4]
```diff
@@ -439,5 +413,95 @@ class RailJsonInfra(BaseModel):
...
     ALL_OBJECT_TYPES.append(t)
 
-if __name__ == "__main__":
-    print(RailJsonInfra.schema_json(indent=2))
+# Extensions
+
+
+def register_extension(object, name):
+    """
+    This decorator is used to register an extension.
```
Comment (flomonster) [inline api/osrd_infra/schemas/infra.py]:
```suggestion
    This decorator is used to easily add an extension to an existing object.
```

File: api/osrd_infra/schemas/infra.py [hunk 5]
```diff
@@ -439,5 +413,95 @@ class RailJsonInfra(BaseModel):
...
+    track_name: constr(max_length=255) = Field(description="Name corresponding to the track used")
+
+
+@register_extension(object=OperationalPoint, name="sncf")
+class OperationalPointSncfExtension(BaseModel):
+    ci: int = Field(description="Infra code from THOR of the corresponding operational point")
+    ch: constr(max_length=2) = Field(description="Side code from THOR of the operational point")
+    ch_short_label: constr(max_length=255)
+    ch_long_label: constr(max_length=255)
+    trigram: constr(max_length=3)
```
Comment (flomonster) [inline api/osrd_infra/schemas/infra.py]: We could add description

File: api/osrd_infra/schemas/infra.py [hunk 6]
```diff
@@ -439,5 +413,95 @@ class RailJsonInfra(BaseModel):
...
+    aspects: List[str]
+    comment: str
+    default_aspect: str = Field(description="Aspect displayed when no train is around")
+    installation_type: str
+    is_in_service: bool = Field(description="Is the signal is in service")
+    is_lightable: bool = Field(description="Is the signal is lightable")
+    is_operational: bool = Field(description="Is the signal is operational")
+    label: str
+    physical_organization_group: str
+    responsible_group: str
```
Comment (multun) [inline api/osrd_infra/schemas/infra.py]: I think these can be safely removed

File: core/src/main/java/fr/sncf/osrd/railjson/schema/infra/trackranges/SingleDirectionalRJSTrackRange.java [hunk 7]
```diff
@@ -1,37 +1,31 @@
...
+import fr.sncf.osrd.railjson.schema.common.graph.EdgeDirection;
+
+@SuppressFBWarnings({"UWF_UNWRITTEN_PUBLIC_OR_PROTECTED_FIELD"})
+public class SingleDirectionalRJSTrackRange extends RJSTrackRange {
+    public EdgeDirection direction;
+    public String track;
+
+    @Override
+    public ApplicableDirection getNavigability() {
+        if (direction == EdgeDirection.START_TO_STOP) return ApplicableDirection.START_TO_STOP;
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/railjson/schema/infra/trackranges/SingleDirectionalRJSTrackRange.java]: Weird checkstyle

**Review by multun — DISMISSED** [review]

**Review by flomonster — CHANGES_REQUESTED** [review]

File: api/osrd_infra/schemas/infra.py (hunk 6 above)
Comment (flomonster) [reply api/osrd_infra/schemas/infra.py]: Done

**Review by flomonster — DISMISSED** [review]

**Review by flomonster — DISMISSED** [review]

**Pull request merged**